src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,314 @@
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        /// <param name="expression">The expression to match with, such as "*.foo".</param>
+        /// <param name="name">The name to check against the expression.</param>
+        /// <param name="ignoreCase">True to ignore case (default).</param>
+        /// <remarks>
+        /// This is based off of System.IO.PatternMatcher used in FileSystemWatcher, which is based off
+        /// of RtlIsNameInExpression, which defines the rules for matching DOS wildcards ('*', '?', '&lt;', '&gt;', '"').
+        /// 
+        /// Like PatternMatcher, matching will not line up with Win32 behavior unless you transform the expression
+        /// using <see cref="TranslateExpression(string)"/>
+        /// </remarks>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+        {
+            string debug = new string(name);
+
+            // The idea behind the algorithm is pretty simple. We keep track of all possible locations
+            // in the regular expression that are matching the name. When the name has been exhausted,
+            // if one of the locations in the expression is also just exhausted, the name is in the
+            // language defined by the regular expression.
+
+            if (string.IsNullOrEmpty(expression) || name.Length == 0)
+                return false;
+
+            if (expression[0] == '*')
+            {
+                // Just * matches everything
+                if (expression.Length == 1)
+                    return true;
+
+                if (expression.IndexOfAny(s_wildcardChars, startIndex: 1) == -1)
+                {
+                    // Handle the special case of a single  starting *, which essentially means "ends with"
+
+                    // If the name doesn't have enough characters to match the remaining expression, it can't be a match.
+                    if (name.Length < expression.Length - 1)
+                        return false;
+
+                    // See if we end with the expression (minus the *, of course)
+                    return name.EndsWithOrdinal(expression.AsReadOnlySpan().Slice(1), ignoreCase);
+                }
+            }
+
+            Span<int> temp;
+
+            int nameOffset = 0;
+            int expressionOffset;
+
+            int priorMatch;
+            int currentMatch;
+            int priorMatchCount;
+            int matchCount = 1;
+
+            char nameChar = '\0';
+            char expressionChar;
+
+            int* priorStack = stackalloc int[16];
+            int* currentStack = stackalloc int[16];
+            Span<int> priorMatches = new Span<int>(priorStack, 16);
+            Span<int> currentMatches = new Span<int>(currentStack, 16);
```

Comment:
Any reason these can't use the recently added C# stackalloc/span support, e.g.

```c#
Span<int> priorMatches = stackalloc int[16];
Span<int> currentMatches = stackalloc int[16];
```

at which point, it looks like this method would no longer need to be `unsafe`.

Reply:
Wasn't aware of it. Tried switching and it wouldn't let me do the swap between the two anymore (at the end of the method).

Reply:
> Tried switching and it wouldn't let me do the swap between the two anymore (at the end of the method).

@JeremyKuhne, what error do you get?
@VSadov?

Reply:
```
"P:\repos\corefx\src\System.IO.FileSystem\src\System.IO.FileSystem.csproj" (default target) (1) ->
(CoreCompile target) ->
  System\IO\DosMatcher.cs(298,24): error CS8352: Cannot use local 'priorMatches' in this context because it may expose referenced variables outside of their declaration scope [P:\repos
\corefx\src\System.IO.FileSystem\src\System.IO.FileSystem.csproj]
```